ftplugin/perl/perl.xpt.vim

```diff
@@ -43,109 +44,261 @@ XPT fornn hidden=1
 
 XPT whilenn hidden=1
 
+" Not perl's way of doing things
+XPT while1 hidden=1
+XPT while0 hidden=1
+
 
 XPT perl " #!/usr/bin/env perl
 #!/usr/bin/env perl
 
-..XPT
+use strict;
+use warnings;
 
 
-XPT xif " .. if ..;
-`expr^ if `cond^;
+..XPT
 
+XPT iif " .. if ..;
+if`$SPcmd^(`$SParg^`condition^`$SParg^)`$BRif^{{
+	`cursor^
+}}
 
-XPT xwhile " .. while ..;
-`expr^ while `cond^;
+XPT wwhen " .. when ..;
+when`$SPcmd^(`$SParg^m/`regex^/`$SParg^)`$BRif^{{
+	`cursor^
+}}
 
+XPT uunless " .. unless ..;
+unless`$SPcmd^(`$SParg^`condition^`$SParg^)`$BRif^{{
+	`cursor^
+}}
 
-XPT xunless " .. unless ..;
-`expr^ unless `cond^;
+XPT sub " sub .. { .. }
+sub `fun_name^`$BRfun^{
+	`cursor^
+}
 
+XPT ssub " sub .. ( .. ) { .. }
+XSET arg*|post=ExpandIfNotEmpty(', ', 'arg*')
+sub `fun_name^(`$SParg^`arg*^`$SParg^)`$BRfun^{
+	`cursor^
+}
 
-XPT xforeach " .. foreach ..;
-`expr^ foreach @`array^;
+XPT psub " sub :prototype( .. ) ( .. ) { .. }
+XSET arg*|post=ExpandIfNotEmpty(', ', 'arg*')
+sub `fun_name^ :prototype(`proto^) (`$SParg^`arg*^`$SParg^)`$BRfun^{
+	`cursor^
+}
 
+XPT asub " sub {}
+sub {
+	`cursor^
+}
 
-XPT sub " sub .. { .. }
-sub `fun_name^`$BRfun^{
-    `cursor^
+XPT assub " sub ( .. ) { .. }
+XSET arg*|post=ExpandIfNotEmpty(', ', 'arg*')
+sub (`arg*^) {
+	`cursor^
 }
```

Comment:
In my experience, using more than one snippet name for several snippet variants is not a very good idea.

It's hard to remember all the names and their functions.

A single snippet `sub` with some extensibility would be my choice.

If you do need more than one snippet for `sub`, having them all starting with `sub` will make it easier to search through the popup menu.

Reply:
Yeah, very nice, I'm changing that

Reply:
```
XPT sub " sub .. { .. }
sub `fun_name^`$BRfun^{
	`cursor^
}

XPT subsigna " sub .. ( .. ) { .. }
XSET arg*|post=ExpandIfNotEmpty(', ', 'arg*')
sub `fun_name^(`$SParg^`arg*^`$SParg^)`$BRfun^{
	`cursor^
}

XPT subprotosigna " sub :prototype( .. ) ( .. ) { .. }
XSET arg*|post=ExpandIfNotEmpty(', ', 'arg*')
sub `fun_name^ :prototype(`proto^) (`$SParg^`arg*^`$SParg^)`$BRfun^{
	`cursor^
}

XPT subanony " sub {}
sub {
	`cursor^
}

XPT subanonysigna " sub ( .. ) { .. }
XSET arg*|post=ExpandIfNotEmpty(', ', 'arg*')
sub (`arg*^) {
	`cursor^
}

```
signa => signatures subroutines
anony => anonymous subroutines
anonysig => anonymous signature subroutines

sounds good?

Reply:
> ```
> XPT sub " sub .. { .. }
> sub `fun_name^`$BRfun^{
> 	`cursor^
> }
> 
> XPT subanony " sub {}
> sub {
> 	`cursor^
> }
> ```

These two are almost the same. IMHO, it can be merged into one.
There is a **cancel** key to clear current placeholder and go to the next:

```
			   *s_<CR>*  *xpt-key-clear* *g:xptemplate_nav_cancel*
								{default:<Cr>}
	Press to clear current place holder and jump to next.
	Use |g:xptemplate_nav_cancel| to change this mapping. Example: >
		let g:xptemplate_nav_cancel = '<C-c>'

```

Thus `sub`, `<trigger-key>`, `<cancel-key>` will just result in:
```
sub {
    // cursor
}
```

> ```
> XPT subsigna " sub .. ( .. ) { .. }
> XSET arg*|post=ExpandIfNotEmpty(', ', 'arg*')
> sub `fun_name^(`$SParg^`arg*^`$SParg^)`$BRfun^{
> 	`cursor^
> }
> 
> XPT subprotosigna " sub :prototype( .. ) ( .. ) { .. }
> XSET arg*|post=ExpandIfNotEmpty(', ', 'arg*')
> sub `fun_name^ :prototype(`proto^) (`$SParg^`arg*^`$SParg^)`$BRfun^{
> 	`cursor^
> }
> 
> XPT subanonysigna " sub ( .. ) { .. }
> XSET arg*|post=ExpandIfNotEmpty(', ', 'arg*')
> sub (`arg*^) {
> 	`cursor^
> }
> ```
> 
> ```
> signa      => signatures subroutines
> anony      => anonymous subroutines
> anonysig  => anonymous signature subroutines
> ```
> 
> sounds good?

The similar to the other snippets: if a placeholder is easy to remove, maybe it is not necessary to create another snippet without that placeholder.

The second thing is that after using xptemplate for a while, I am getting unsure about the necessity of argument repetition, such as this: `XSET arg*|post=ExpandIfNotEmpty(', ', 'arg*')`.
In my vim I mapped `,` to `,<space>`, so I only need one keystroke `,` to start editing the next argument.


With this snippet:
```
XPT sub2 " sub ( .. ) { .. }
XSET arg*|post=ExpandInsideEdge(', ', '')
sub `fun_name^`(`arg*`)^ {
	`cursor^
}
```
`sub2`, `<trigger-key>`, `<cancel-key>`, `<cancel-key>` you will get:
```
sub {
    // cursor
}
```

`sub2`, `<trigger-key>`, `<cancel-key>`, `foo`, `<nav-next>`, `bar`,  you will get:
```
sub (foo, bar) {
    // cursor
}
```

`sub2`, `<trigger-key>`, `wow`, `<cancel-key>` you will get:
```
sub wow {
    // cursor
}
```

I believe the these default snippets should be common and general. :)

Reply:
Wow :)

That works though we have an extra space in between `sub` and `{` in the case of `sub { ... }`

This is yours
```
XPT sub2 " sub ( .. ) { .. }
XSET arg*|post=ExpandInsideEdge(', ', '')
sub `fun_name^`(`arg*`)^ {
	`cursor^
}
```
So in this your snippet, how can I insert `$SParg` beside the corners of `()`?

```
XPT subp " sub :prototype\( .. ) \( .. ) { .. }
XSET arg*|post=ExpandIfNotEmpty(', ', 'arg*')
sub `fun_name^ :prototype(`proto^) (`$SParg^`arg*^`$SParg^)`$BRfun^{
	`cursor^
}
```
And for this one I suggest we should not change anything to avoid headache

Reply:
> Wow :)
> 
> That works though we have an extra space in between `sub` and `{` in the case of `sub { ... }`
> 
> This is yours
> 
> ```
> XPT sub2 " sub ( .. ) { .. }
> XSET arg*|post=ExpandInsideEdge(', ', '')
> sub `fun_name^`(`arg*`)^ {
> 	`cursor^
> }
> ```
> 
> So in this your snippet, how can I insert `$SParg` beside the corners of `()`?

No, I did not figure out a way to do that:(.

> ```
> XPT subp " sub :prototype\( .. ) \( .. ) { .. }
> XSET arg*|post=ExpandIfNotEmpty(', ', 'arg*')
> sub `fun_name^ :prototype(`proto^) (`$SParg^`arg*^`$SParg^)`$BRfun^{
> 	`cursor^
> }
> ```
> 
> And for this one I suggest we should not change anything to avoid headache

Alright

Reply:
> I too did not see the necessity, just took that from `python.xpt.vim` and IMO it slows down, would remove it if you want me to do that.

Less is better:D

Reply:
Sorry for the delay :(

Updated the repo, please check if everything is fine.

Reply:
> Sorry for the delay :(
> 
> Updated the repo, please check if everything is fine.

Looks good to me. Let me merge it.
(It requires a manually merge:()

Reply:
Hmmm, that's strange

`This branch has no conflicts with the base branch`

Reply:
Well...I maintain the git history of this repo in a kind of weird way. I tried to keep every feature, every test case, and every snippet in a branch. And merge them all into one branch `mix`. Every change should be applied to its belonging branch. 

When publishing, build `master` branch from `mix`. This is my git history:

<img width="1438" alt="Screen Shot 2022-09-08 at 15 47 19" src="https://user-images.githubusercontent.com/44069/189065750-03a940dd-57df-41d7-9451-cc4e13f2c868.png">